airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py

```diff
@@ -0,0 +1,445 @@
+                return False
+        if self._url is not None and prepared_request.url is not None:
+            url_without_params = prepared_request.url.split("?")[0]
+            if url_without_params != self._url:
+                return False
+        if self._params is not None:
+            parsed_url = parse.urlsplit(prepared_request.url)
+            params = dict(parse.parse_qsl(str(parsed_url.query)))
+            if not self._match_dict(params, self._params):
+                return False
+        if self._headers is not None:
+            if not self._match_dict(prepared_request.headers, self._headers):
+                return False
+        return True
+
+
+class AbstractAPIBudget(abc.ABC):
+    """Interface to some API where client allowed to have N calls per T interval.
+
+    Important: APIBudget is not doing any API calls, the end user code is responsible to call this interface
+        to respect call rate limitation of the API.
+
+    It supports multiple policies applied to different group of requests. To distinct these groups we use RequestMatchers.
+    Individual policy represented by MovingWindowCallRatePolicy and currently supports only moving window strategy.
+    """
+
+    @abc.abstractmethod
+    def add_policy(self, request_matcher: RequestMatcher, policy: AbstractCallRatePolicy) -> None:
+        """Add policy for calls
+
+        :param request_matcher: callable to match request object with corresponding policy
+        :param policy: to acquire calls
+        :return:
+        """
+
+    @abc.abstractmethod
+    def acquire_call(self, request: Any, block: bool = True, timeout: Optional[float] = None) -> None:
+        """Try to get a call from budget, will block by default
+
+        :param request:
+        :param block: when true (default) will block the current thread until call credit is available
+        :param timeout: if set will limit maximum time in block, otherwise will wait until credit is available
+        :raises: CallRateLimitHit - when no credits left and if timeout was set the waiting time exceed the timeout
+        """
+
+    @abc.abstractmethod
+    def match_request(self, request: Any) -> Optional[AbstractCallRatePolicy]:
+        """Find matching call rate policy for specific request"""
+
+    @abc.abstractmethod
+    def update_from_response(self, request: Any, response: Any) -> None:
+        """Update budget information based on response from API
+
+        :param request: the initial request that triggered this response
+        :param response: response from the API
+        """
+
+
+class APIBudget(AbstractAPIBudget):
+    """Default APIBudget implementation"""
+
+    maximum_attempts_to_acquire: int = 10
+
+    def __init__(self) -> None:
+        """Constructor"""
+        self._policies: list[tuple[RequestMatcher, AbstractCallRatePolicy]] = []
+
+    def add_policy(self, request_matcher: RequestMatcher, policy: AbstractCallRatePolicy) -> None:
```

Comment:
is it safe to add policies at runtime? If not, can we initialize the APIBudget with its policies?